### ISSUE TITLE
Chunks of a table go 'missing' under certain circumstances

### ISSUE THREAD

**Issue by drfraser** [timescale/timescaledb#3601]
**Relevant system information:**
 - OS: Ubuntu 20
 - PostgreSQL version 13.4
 - TimescaleDB version 2.4.1
 - Installation method: apt install

**Describe the bug**
A select statement to list the unique dates in a table revealed only the last chunk was being searched, not the first chunk

**To Reproduce**
1. Create a hypertable with a date (not datetime column) (and other columns) and an interval of 50 years
2. Insert data that goes from 1993-01-01 to 2020-12-31 with a gap of all of 2017
3. after insertion, compress the table using 
4. SELECT compress_chunk(i) from show_chunks('{tablename}', older_than => date(now() + interval '1 week')) i;
5. But that does not compress the very last chunk, so compress it with an explicit SELECT compress_chunk()
6. select distinct dateof from sy05_ta_0002m_adm0_abs_01m_0_05;

**Expected behavior**
All of the dates in the table from 1993 to 2020 should have been returned

**Actual behavior**
Only the dates in the last chunk of the table were returned (2019 to 2020)

**Additional context**
Only data associated with each day or at most one month are being inserted into the table - time is not relevant for our application. So the interval of 50 years was chosen to have the first chunk stretch from 1970-01-01 to 2020-01-01 But the enddate of the first chunk is some time in 2019, which is fine.

For various reasons, I wanted to force the compression of data upon insertion because it won't be updated very much. So the insert script was running item 4 above to compress the table, then a manual SELECT compress_chunk was at the end to forcibly compress anything uncompressed (used in other circumstances).  So the two chunks of the table were compressed as I'd expect.

But in trying to retrieve data from the table, only data from the chunk starting in 2019-05-01 was returned. I tried uncompressing the first chunk and then both but still it was ignored. Whatever index item 6 uses seemed to only have the extent of the dates associated with the table to be 2019 to 2020

So I deleted everything and made the interval 100 years. With one chunk, compressed with item 5, things worked.

So I made the interval 50 years and stopped the compression in item 4 - now the two chunks of the table are compressed together one after another (item 5).  This time, the select distinct dates SQL worked. No good idea what is going on, but it is repeatable.

I have not had this trouble with data where there are no gaps in it. I can keep the interval at 100 years, but am going to have to validate the backfill process pretty heavily.

**Comment by drfraser** [comment]
need-more-info: what exactly?

Table structure and related SQL:
```
CREATE TABLE {schema}.{prefix}{table} (
  dateof DATE NOT NULL,
  leadtime SMALLINT NOT NULL,
  geocode CHAR(2) NOT NULL,
  value real NULL
);
CREATE INDEX on {schema}.{prefix}{table} (geocode, leadtime, dateof desc);
SELECT create_hypertable('{table}','dateof', chunk_time_interval => INTERVAL '50 year', create_default_indexes => FALSE);
ALTER TABLE {schema}.{prefix}{table} SET (timescaledb.compress, timescaledb.compress_segmentby = 'geocode, leadtime',
                                 timescaledb.compress_orderby = 'dateof DESC');
SELECT add_compression_policy('{table}', INTERVAL '1 day');
SELECT add_retention_policy('{table}', INTERVAL '100 years');
```
The data being inserted is a CSV file of 1 row per dateof/leadtime/geocode, and the date range is from 1993-01-01 to 2020-12-01 with the year of 2017 missing entirely.  so two chunks of 50 years each get created.

If the insert script does not immediately compress the hunks after inserting the data, then things are fine.  But we are using Timescale so the data will be compressed, so ...

If the chunks are compressed right after data insertion, then it is like the first chunk doesn't even exist and selecting the distinct dates in the table only return 2019-01-01 to 2020-12-01. But I can see the chunk in the internal Timescale tables.
The count of all records in the table also returns a number far smaller than what it ought to be.

If you need internal Timescale things, then let me know the queries to run

**Comment by svenklemm** [comment]
I tried to follow your instructions and i wasnt able to reproduce it.

```
dev=# select count(*) from i3601 ;
 count
-------
  9862
(1 row)

Time: 21338.351 ms (00:21.338)
dev=# select count(*) FROM (select distinct d from i3601)s;
 count
-------
  9862
(1 row)
```

Ideally you would provide a minimal script that creates an environment that exhibits the erraneous behaviour.

**Comment by drfraser** [comment]
There are 444000+ rows in the table, so I won't make an example that uses all that data.... what is annoying is the inconsistency - I can't make it happen all the time. But I will work on this, though I need to get this MVP working ASAP...

I was using "select count(distinct dateof) from table" - I will add the output of EXPLAIN on my next update

**Comment by drfraser** [comment]
Hi, it has happened again and I investigated things a bit.  I have attached the source file of data as it's not too big. Also attached is a sort of writeup of various SQL statements I ran - the first chunk is definitely not getting processed. It is mentioned in the query plan, but the row count is always 0

FYI: the geocode column is now CHAR(4) but everything else as above is the same

I won't delete this table, so let me know what detailed SQL I ought to run to gather info.

[bad_data.csv.gz](https://github.com/timescale/timescaledb/files/7212764/bad_data.csv.gz)

[writeup.txt](https://github.com/timescale/timescaledb/files/7212858/writeup.txt)


this is definitely a bit concerning - there are NO gaps in this data in terms of time (except for the yearly step) so I guess that really isn't a factor

**Comment by svenklemm** [comment]
I tried to reproduce but i could not. Please provide a single script that has all the setup included including data, that when run in a fresh database exhibits the behaviour. The scripts you provided seem to be inconsistent since your table defnition has 4 columns but your csv only has 3. Also please try to make all the scripts runnable in plain psql since you seem to have variables in them which makes running them harder.

**Comment by svenklemm** [comment]
Are you sure you are not deleting the data yourself with your retention policy? Can you reproduce this all without the retention policy? I can see from your queries that the 1st chunk has no data.
What is the result of the following query.
```explain select tableoid::regclass,count(*) from <table> group by tableoid;```

**Comment by drfraser** [comment]
Hi,

thanks for taking a look at this. I haven't had a chance to look into this since I filed the bug - I needed to get work done for the client, it only ever occurred once more and I needed to get things into production...   Then I had to deal with another client.

I have been meaning to write a script that repeatedly does setup / insert / teardown because there just wasn't any consistency in when it would occur.  I don't remember details right now, but I got the impression the 1st chunk did have data in it, based on the investigation I did - but the query plan was going wrong. 

And no, the retention policy is set to 100 years, so I am sure it wasn't getting deleted.  This use case is actually not the norm for Timescale - we are dealing with archiving climate data (which is a time series) and which will never get deleted. The earliest data begins in 1950. The compression is turning out to be very useful (lots of disk space saved) so the client is happy with Timescale.

Once I figure out how to reproduce or at least nail down the context around when it happens, I will update this issue.

**Comment by drfraser** [comment]
Attached is  a ZIP of the test script I made, the dataset and a writeup of the SQL used to examine the bad table.

[ts_script.zip](https://github.com/timescale/timescaledb/files/7617772/ts_script.zip)

The script should be straightforward to use after a bit of configuring, e.g. IP address of Timescale server. See the comments about the psql to setup the database

TABLE_NAME is a parameter so broken tables could be kept instead of being overwritten

I added the optional '-s' command line switch to test if delays might be a factor, which they are.

**Relevant system information:**
 - Server OS: Ubuntu 20 (running in a Parallels VM)
 - PostgreSQL version 13.4
 - TimescaleDB version 2.4.1 (and 2.5.2)
 - Client program OS: OS X 10.11 (and also Ubuntu 20)
 - iMac 2010, 3.2GHz Core i3
 - psycopg2 versions:  psycopg2-binary==2.8.6, psycopg2-binary==2.9.2, psycopg2==2.8.6, psycopg2==2.9.2
 - Python 3.8.10
 
When everything (client program and server) was local to my machine, the script would always result in the same problem of the first chunk never 'having' any data and so count(*) etc fails to return the correct results.

this is obviously an excessive stress test - changing the chunk time interval from 1 year to 100, 1 at a time, is unrealistic, but it does flush out the problem. There was never any consistency as to the chunk time interval when things broke except that when artificial delays were imposed with '-s', the final chunk interval (when the script halted) was between 30 and 50, not under 10. 

I did experience this problem several times with a realistic data set and realistic chunk interval, so something is off....

But when I switched to a local client (OS X) and the database being on my server in the US (Debian 9, Timescale 2.4.1), there were no problems, even when there were no artificial delays imposed with sleep().  So my impression is that when everything is local, Timescale gets "overwhelmed" in some sense and something breaks with respect to compression of the first chunk. The delay imposed by communicating with with my Debian server over the Internet added enough delay to things so the problem never occurred. I have not configured Timescale beyond running timescaledb_tune.

**Comment by drfraser** [comment]
I decided to determine if python/psycopg2  was perhaps the issue.  Attached is a GZIP (NOT zip) with a psql-only script that reliably reproduces the problem. Also included are:

[missing_chunk.zip](https://github.com/timescale/timescaledb/files/7755584/missing_chunk.zip)

missing_chunk.sql - the script that creates the bad table and a good table
missing_chunk.out - the output of the script
missing_explains.txt - the explain analyzes used to show the query plans
missing.txt - the sql used to track down the problem

"***" highlights my comments

TLDR; the first uncompressed chunk of the bad table is associated with two compressed chunks (incorrectly) according to table 'compression_chunk_size' and so the query plan/result of select breaks because the second compressed chunk has no data in it.  And the first compressed chunk thus gets "lost"

**Comment by svenklemm** [comment]
This should be fixed by https://github.com/timescale/timescaledb/pull/4482
Feel free to reopen if you can still reproduce this with 2.7.1.
